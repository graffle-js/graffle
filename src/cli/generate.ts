#!/usr/bin/env node

import { Command } from '@molt/command'
import * as Path from 'node:path'
import { z } from 'zod'
import { generateFiles } from '../layers/2_generator/files.js'
import { urlParseSafe } from '../lib/prelude.js'

const args = Command.create().description(`Generate a type safe GraphQL client.`)
  .parameter(
    `name`,
    z.string().min(1).optional().describe(
      `The name of your client. If you are not generating multiple clients you probably do not need this. Otherwise you need to differentiate the clients so that their global type registrations do not conflict. It is possible to leave one client unnamed which will become the default client at the type level (e.g. in configuration etc.)`,
    ),
  )
  .parameter(
    `schema`,
    z.string().min(1).describe(
      `Path to where your GraphQL schema is. If a URL is given it will be introspected. Otherwise assumed to be a file path to your GraphQL SDL file.`,
    ),
  )
  .parametersExclusive(
    `schemaErrorType`,
    $ =>
      $.parameter(
        `schemaErrorTypes`,
        z.boolean().describe(
          `Use the schema error types pattern. All object types whose name starts with "Error" will be considered to be error types. If you want to specify a custom name pattern then use the other parameter "schemaErrorTypePattern".`,
        ),
      )
        .parameter(
          `schemaErrorTypePattern`,
          z.string().min(1).describe(
            `Designate objects whose name matches this JS regular expression as being error types in your schema.`,
          ),
        ).default(`schemaErrorTypes`, true),
  )
  .parameter(
    `defaultSchemaUrl`,
    z.union([
      z.boolean().describe(
        `If a GraphQL endpoint is given for "--schema", should it be set as the default URL in the generated client. If true then the client will default to using this URL when sending requests.`,
      ),
      z.string().min(1).describe(
        `A GraphQL endpoint to be used as the default URL in the generated client for requests.`,
      ),
    ]).default(true),
  )
  .parameter(
    `output`,
    z.string().min(1).default(`./graffle`).describe(
      `Directory path for where to output the generated TypeScript files.`,
    ),
  )
  .parameter(`format`, z.boolean().describe(`Format the generated files using dprint.`).default(true))
  .parameter(
    `libraryPathClient`,
    z.string().optional().describe(
      `Custom location for where the generated code should import the Graffle "client" module from.`,
    ),
  )
  .parameter(
    `libraryPathSchema`,
    z.string().optional().describe(
      `Custom location for where the generated code should import the Graffle "schema" module from.`,
    ),
  )
  .parameter(
    `libraryPathScalars`,
    z.string().optional().describe(
      `Custom location for where the generated code should import the Graffle "scalars" module from.`,
    ),
  )
  .settings({
    parameters: {
      environment: false,
    },
  })
  .parse()

const url = urlParseSafe(args.schema)
const defaultSchemaUrl = typeof args.defaultSchemaUrl === `string`
  ? new URL(args.defaultSchemaUrl)
  : args.defaultSchemaUrl

await generateFiles({
  sourceSchema: url
    ? { type: `url`, url }
    : { type: `sdl`, dirPath: Path.dirname(args.schema) },
  defaultSchemaUrl,
  name: args.name,
  libraryPaths: {
    client: args.libraryPathClient,
    schema: args.libraryPathSchema,
    scalars: args.libraryPathScalars,
  },
  outputDirPath: args.output,
  format: args.format,
  errorTypeNamePattern: args.schemaErrorType._tag === `schemaErrorTypePattern`
    ? new RegExp(args.schemaErrorType.value)
    : args.schemaErrorType.value
    ? /^Error.+/
    : undefined,
})
