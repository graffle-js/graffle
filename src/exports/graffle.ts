export { Transport } from '../context/fragments/transports/dataType/_namespace.js'
export * as Scalars from './_Scalars.js'
export * from './presets/minimal_.js'
