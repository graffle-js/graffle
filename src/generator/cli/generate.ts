#!/usr/bin/env node

import { Command } from '@molt/command'
import * as Path from 'node:path'
import { z } from 'zod'
import { toAbsolutePath } from '../../lib/fsp.js'
import { isError, urlParseSafe } from '../../lib/prelude.js'
import { Generator } from '../__.js'
import { type Input, OutputCase } from '../config/input.js'

const args = Command.create().description(`Generate a type safe GraphQL client.`)
  .parameter(
    `name`,
    z.string().min(1).optional().describe(
      `The name of your client. If you are not generating multiple clients you probably do not need this. Otherwise you need to differentiate the clients so that their global type registrations do not conflict. It is possible to leave one client unnamed which will become the default client at the type level (e.g. in configuration etc.)`,
    ),
  )
  .parameter(
    `schema`,
    z.string().min(1).optional().describe(
      `Path to where your GraphQL schema is. If a URL is given it will be introspected. Otherwise assumed to be a path to your GraphQL SDL file. If a directory path is given, then will look for a "schema.graphql" within that directory. Otherwise will attempt to load the exact file path given. If omitted, then your project must have a configuration file which supplies the schema source.`,
    ),
  )
  .parameter(
    `project`,
    z.string().optional().describe(
      `Path to your configuration file. By default will look for "graffle.config.{ts,js,mjs,mts}" in the current working directory. If a directory path is given, then will look for "graffle.config.{ts,js,mjs,mts}" in that directory.`,
    ),
  )
  .parameter(
    `defaultSchemaUrl`,
    z.union([
      z.boolean().describe(
        `If a GraphQL endpoint is given for "--schema", should it be set as the default URL in the generated client. If true then the client will default to using this URL when sending requests.`,
      ),
      z.string().min(1).describe(
        `A GraphQL endpoint to be used as the default URL in the generated client for requests.`,
      ),
    ]).optional(),
  )
  .parameter(
    `output`,
    z.string().min(1).optional().describe(
      `Directory path for where to output the generated TypeScript files. By default will be './graffle' in the project root.`,
    ),
  )
  .parameter(
    `outputCase`,
    z.nativeEnum(OutputCase).default(`kebab`).describe(
      `The case format to use for the generated file names.`,
    ),
  )
  .parameter(
    `format`,
    z.boolean().describe(
      `Try to format the generated files. At attempt to use dprint will be made. You need to have these dependencies installed in your project: @dprint/formatter, @dprint/typescript.`,
    )
      .optional(),
  )
  .settings({
    parameters: {
      environment: false,
    },
  })
  .parse()

// --- Resolve Config File ---

const configModule = await Generator.Config.load({ filePath: args.project })
if (isError(configModule)) throw configModule
if (!configModule.builder && args.project) {
  throw new Error(`Could not find a configuration file at "${configModule.paths.join(`, `)}".`)
}

if (configModule.builder) {
  // todo: nice logging, levels etc.
  console.log(`Using file config found at "${configModule.path}".`)
}

// --- Resolve Default Schema URL ---

const defaultSchemaUrl = typeof args.defaultSchemaUrl === `string`
  ? new URL(args.defaultSchemaUrl)
  : args.defaultSchemaUrl

// --- Resolve Schema ---

const url = args.schema ? urlParseSafe(args.schema) : null

const schemaViaCLI = args.schema
  ? url
    ? { type: `url` as const, url }
    : { type: `sdlFile` as const, dirOrFilePath: Path.join(process.cwd(), args.schema) }
  : undefined

if (schemaViaCLI && configModule.builder?._.input.schema) {
  console.log(`WARNING: Overriding config file schema configuration with command line input.`)
}

const schema = schemaViaCLI ?? configModule.builder?._.input.schema

if (!schema) {
  throw new Error(`No schema source provided. Either specify a schema source in the config file or via the CLI.`)
}

const currentWorkingDirectory = configModule.path ? Path.dirname(configModule.path) : process.cwd()

// --- Merge Inputs ---

const input: Input = {
  ...configModule.builder?._.input,
  currentWorkingDirectory,
  schema,
  outputCase: args.outputCase,
}

if (defaultSchemaUrl !== undefined) input.defaultSchemaUrl = defaultSchemaUrl
if (args.format !== undefined) input.format = args.format
if (args.name !== undefined) input.name = args.name
if (args.output !== undefined) input.outputDirPath = toAbsolutePath(process.cwd(), args.output)

// --- Generate ---

await Generator.generate(input)
