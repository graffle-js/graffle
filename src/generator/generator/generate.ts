import { createConfig } from '../config/config.js'
import type { ConfigInit } from '../config/configInit.js'
import { ModuleGenerator_ } from '../generators/_.js'
import { ModuleGenerator__ } from '../generators/__.js'
import { ModuleGeneratorClient } from '../generators/Client.js'
import { ModuleGeneratorData } from '../generators/Data.js'
import { ModuleGeneratorGlobal } from '../generators/global.js'
import { ModuleGeneratorMethodsDocument } from '../generators/MethodsDocument.js'
import { ModuleGeneratorMethodsRoot } from '../generators/MethodsRoot.js'
import { ModuleGeneratorMethodsSelect } from '../generators/MethodsSelect.js'
import { ModuleGeneratorScalar } from '../generators/Scalar.js'
import { ModuleGeneratorSchema } from '../generators/Schema.js'
import { ModuleGeneratorSchemaDrivenDataMap } from '../generators/SchemaDrivenDataMap.js'
import { ModuleGeneratorSelect } from '../generators/Select.js'
import { ModuleGeneratorSelectionSets } from '../generators/SelectionSets.js'
import { getFileName, isExportsModule } from '../helpers/moduleGenerator.js'

const moduleGenerators = [
  ModuleGeneratorGlobal,
  ModuleGeneratorClient,
  ModuleGeneratorData,
  ModuleGeneratorScalar,
  // Packaging Stuff
  ModuleGenerator__,
  ModuleGenerator_,
  // Schema Stuff
  ModuleGeneratorSchema,
  ModuleGeneratorSchemaDrivenDataMap,
  // Interface Stuff
  ModuleGeneratorSelectionSets,
  ModuleGeneratorSelect,
  ModuleGeneratorMethodsSelect,
  ModuleGeneratorMethodsRoot,
  ModuleGeneratorMethodsDocument,
]

export const generate = async (input: ConfigInit) => {
  const config = await createConfig(input)

  const generatedModules = await Promise.all(
    moduleGenerators
      .map(generator => generator.generate(config))
      .map(async code => ({
        ...code,
        content: await config.formatter.formatText(code.content),
      })),
  )

  if (config.paths.project.outputs.sdl && config.schema.via !== `sdl`) {
    await config.fs.writeFile(config.paths.project.outputs.sdl, config.schema.sdl)
  }

  // todo clear directory before generating so that removed or renamed files are cleaned up.
  await config.fs.mkdir(config.paths.project.outputs.root, { recursive: true })
  await config.fs.mkdir(config.paths.project.outputs.modules, { recursive: true })

  await Promise.all(
    generatedModules.map((generatedModule) => {
      // dprint-ignore
      const filePath = `${config.paths.project.outputs.root}/${isExportsModule(generatedModule.name) ? `` : `modules/`}${getFileName(config, generatedModule)}`
      return config.fs.writeFile(filePath, generatedModule.content)
    }),
  )
}
