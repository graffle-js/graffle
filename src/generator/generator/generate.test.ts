import { globby } from 'globby'
import * as Memfs from 'memfs'
import { readFile } from 'node:fs/promises'
import * as Path from 'node:path'
import { describe, expect, test } from 'vitest'
import type { ConfigInitSchemaSdl } from '../_.js'
import { generate } from './generate.js'

const schema: ConfigInitSchemaSdl = {
  type: `sdl`,
  sdl: `type Query { ok: Boolean }`,
}

describe(`importFormat`, () => {
  test(`default is jsExtension`, async () => {
    await generate({
      fs: Memfs.fs.promises as any,
      schema,
    })
    const SchemaTs = Memfs.fs.readFileSync(`./graffle/modules/schema.ts`, `utf8`)
    expect(SchemaTs).toMatch(/import.*".\/data.js"/)
  })
  test(`noExtension`, async () => {
    await generate({
      fs: Memfs.fs.promises as any,
      schema,
      importFormat: `noExtension`,
    })
    const SchemaTs = Memfs.fs.readFileSync(`./graffle/modules/schema.ts`, `utf8`)
    expect(SchemaTs).toMatch(/import.*".\/data"/)
  })
})

test(`kitchen-sink generated modules`, async () => {
  const basePath = `./tests/_/schemas/kitchen-sink/graffle`
  const filePaths = await globby(`${basePath}/**/*.ts`)
  for (const filePath of filePaths) {
    const relativeFilePath = Path.relative(basePath, filePath)
    const content = await readFile(filePath, `utf8`)
    expect(content).toMatchSnapshot(relativeFilePath)
  }
})

test(`root-types-mapped`, async () => {
  await Memfs.fs.promises.mkdir(process.cwd(), { recursive: true })
  await generate({
    fs: Memfs.fs.promises as any,
    schema: {
      type: `sdl`,
      sdl: `
        schema {
          query: QueryRoot
        }
        type QueryRoot {
          id: ID
        }
      `,
    },
  })

  const SchemaTs = Memfs.fs.readFileSync(`./graffle/modules/schema.ts`, `utf8`)
  expect(SchemaTs).includes(`operationsAvailable: ["query"]`)
  expect(SchemaTs).includes(`RootUnion: Schema.QueryRoot`)
  expect(SchemaTs).toMatchSnapshot()

  const MethodsRootTs = Memfs.fs.readFileSync(`./graffle/modules/methods-root.ts`, `utf8`)
  expect(MethodsRootTs).includes(`__typename: "QueryRoot"`)
  expect(MethodsRootTs).includes(`InferResult.OperationQuery<`)
  expect(MethodsRootTs).toMatchSnapshot()
})

test(`custom scalars module results in client prefilling those custom scalars`, async () => {
  await Memfs.fs.promises.mkdir(process.cwd(), { recursive: true })
  await Memfs.fs.promises.writeFile(
    `./scalars.ts`,
    `
      import { Graffle } from 'graffle'
      export const Date = Graffle.Scalar.create(\`Date\`, {
        decode: (value) => new globalThis.Date(value),
        encode: (value) => value.toISOString(),
      })
    `,
  )
  await generate({
    fs: Memfs.fs.promises as any,
    schema: {
      type: `sdl`,
      sdl: `
        scalar Date
        type Query {
          date: Date
        }
      `,
    },
  })
  const ScalarTs = Memfs.fs.readFileSync(`./graffle/modules/scalar.ts`, `utf8`)
  expect(ScalarTs).toMatchSnapshot()
})
