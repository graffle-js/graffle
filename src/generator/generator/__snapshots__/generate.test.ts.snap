// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`custom scalars module results in client prefilling those custom scalars 1`] = `
"import type * as $$Utilities from "graffle/utilities-for-generated";

import * as CustomScalars from "../../scalars.js";

export * from "../../scalars.js";
export { Date } from "../../scalars.js";
//
//
//
//
// CUSTOM SCALAR TYPE
// DATE
// --------------------------------------------------------------------------------------------------
//                                                Date
// --------------------------------------------------------------------------------------------------
//
//

export type Date = typeof CustomScalars.Date;
// Without this we get error:
// "Exported type alias 'DateDecoded' has or is using private name 'Date'."
type Date_ = typeof CustomScalars.Date;
export type DateDecoded = $$Utilities.Schema.Scalar.GetDecoded<Date_>;
export type DateEncoded = $$Utilities.Schema.Scalar.GetEncoded<Date_>;

export * from "graffle/generator-helpers/standard-scalar-types";

//
//
//
//
//
//
// ==================================================================================================
//                                              Registry
// ==================================================================================================
//
//
//
//
//
//

export const $registry = {
  map: {
    Date: Date,
  },
} as $Registry;

export type $Registry = $$Utilities.Schema.Scalar.Registry<
  {
    Date: Date_;
  },
  $$Utilities.Schema.Scalar.GetEncoded<Date_>,
  $$Utilities.Schema.Scalar.GetDecoded<Date_>
>;
"
`;

exports[`root-types-mapped 1`] = `
"import * as $$Data from "./data.js";
import * as $$Scalar from "./scalar.js";
import type { Schema as $ } from "graffle/utilities-for-generated";
import type * as $$Utilities from "graffle/utilities-for-generated";

export namespace Schema {
  //
  //
  //
  //
  //
  //
  // ==================================================================================================
  //                                                Root
  // ==================================================================================================
  //
  //
  //
  //
  //
  //

  //                                     RootTypeCustomNameForQuery
  // --------------------------------------------------------------------------------------------------
  //

  export interface RootTypeCustomNameForQuery {
    kind: "Object";
    name: "RootTypeCustomNameForQuery";
    fields: {
      __typename: RootTypeCustomNameForQuery.__typename;
      id: RootTypeCustomNameForQuery.id;
    };
  }

  export namespace RootTypeCustomNameForQuery {
    export interface __typename {
      kind: "OutputField";
      name: "__typename";
      arguments: {};
      inlineType: [1];
      namedType: {
        kind: "__typename";
        value: "RootTypeCustomNameForQuery";
      };
    }

    export interface id {
      kind: "OutputField";
      name: "id";
      arguments: {};
      inlineType: [0];
      namedType: $$NamedTypes.$$ID;
    }
  }

  //
  //
  //
  //
  //
  //
  // ==================================================================================================
  //                                            OutputObject
  // ==================================================================================================
  //
  //
  //
  //
  //
  //

  //
  //
  //
  //
  //
  //
  // ==================================================================================================
  //                                            InputObject
  // ==================================================================================================
  //
  //
  //
  //
  //
  //

  //
  //
  //
  //
  //
  //
  // ==================================================================================================
  //                                             Interface
  // ==================================================================================================
  //
  //
  //
  //
  //
  //

  //
  //
  //
  //
  //
  //
  // ==================================================================================================
  //                                               Union
  // ==================================================================================================
  //
  //
  //
  //
  //
  //

  //
  //
  //
  //
  //
  //
  // ==================================================================================================
  //                                                Enum
  // ==================================================================================================
  //
  //
  //
  //
  //
  //

  //
  //
  //
  //
  //
  //
  // ==================================================================================================
  //                                            ScalarCustom
  // ==================================================================================================
  //
  //
  //
  //
  //
  //

  //
  //
  //
  //
  //
  //
  // ==================================================================================================
  //                                           ScalarStandard
  // ==================================================================================================
  //
  //
  //
  //
  //
  //

  //                                                 ID
  // --------------------------------------------------------------------------------------------------
  //

  export type ID = $.StandardTypes.ID;

  //                                              Boolean
  // --------------------------------------------------------------------------------------------------
  //

  export type Boolean = $.StandardTypes.Boolean;

  //                                               String
  // --------------------------------------------------------------------------------------------------
  //

  export type String = $.StandardTypes.String;

  //
  //
  //
  //
  //
  //
  // ==================================================================================================
  //                                         Named Types Index
  // ==================================================================================================
  //
  //
  //
  //
  //
  //

  /**
   * [1] These definitions serve to allow field selection interfaces to extend their respective object type without
   *     name clashing between the field name and the object name.
   *
   *     For example imagine \`Query.Foo\` field with type also called \`Foo\`. Our generated interfaces for each field
   *     would end up with an error of \`export interface Foo extends Foo ...\`
   */

  namespace $$NamedTypes {
    export type $$RootTypeCustomNameForQuery = RootTypeCustomNameForQuery;
    export type $$ID = ID;
    export type $$Boolean = Boolean;
    export type $$String = String;
  }
}

//
//
//
//
//
//
// ==================================================================================================
//                                               Schema
// ==================================================================================================
//
//
//
//
//
//

export interface Schema<
  $Scalars extends $$Utilities.Schema.Scalar.Registry = $$Scalar.$Registry,
> {
  name: $$Data.Name;
  operationsAvailable: ["query"];
  RootUnion: Schema.RootTypeCustomNameForQuery;
  Root: {
    query: Schema.RootTypeCustomNameForQuery;
    mutation: null;
    subscription: null;
  };
  allTypes: {
    RootTypeCustomNameForQuery: Schema.RootTypeCustomNameForQuery;
  };
  objects: {};
  unions: {};
  interfaces: {};
  scalarNamesUnion: "ID" | "Boolean" | "String";
  scalars: {
    ID: Schema.ID;
    Boolean: Schema.Boolean;
    String: Schema.String;
  };
  scalarRegistry: $Scalars;
  extensions: $$Utilities.GlobalRegistry.TypeExtensions;
}
"
`;

exports[`root-types-mapped 2`] = `
"import type * as $$SelectionSets from "./selection-sets.js";
import type * as $$Schema from "./schema.js";
import type * as $$Utilities from "graffle/utilities-for-generated";

export interface RootTypeCustomNameForQueryMethods<
  $Context extends $$Utilities.Context,
> {
  $batch: $$Utilities.GraffleKit.Context.Configuration.Check.Preflight<
    $Context,
    <$SelectionSet>(
      selectionSet: $$Utilities.Exact<
        $SelectionSet,
        $$SelectionSets.RootTypeCustomNameForQuery<$Context["scalars"]>
      >,
    ) => Promise<
      (null | {}) &
        $$Utilities.HandleOutput<
          $Context,
          $$Utilities.DocumentBuilderKit.InferResult.OperationQuery<
            $$Utilities.AssertExtendsObject<$SelectionSet>,
            $$Schema.Schema<$Context["scalars"]>
          >
        >
    >
  >;
  __typename: $$Utilities.GraffleKit.Context.Configuration.Check.Preflight<
    $Context,
    () => Promise<
      (null | {}) &
        $$Utilities.HandleOutputDocumentBuilderRootField<
          $Context,
          { __typename: "RootTypeCustomNameForQuery" },
          "__typename"
        >
    >
  >;

  id: $$Utilities.GraffleKit.Context.Configuration.Check.Preflight<
    $Context,
    <$SelectionSet>(
      selectionSet?: $$Utilities.Exact<
        $SelectionSet,
        $$SelectionSets.RootTypeCustomNameForQuery.id<$Context["scalars"]>
      >,
    ) => Promise<
      (null | {}) &
        $$Utilities.HandleOutputDocumentBuilderRootField<
          $Context,
          $$Utilities.DocumentBuilderKit.InferResult.OperationQuery<
            { id: $SelectionSet },
            $$Schema.Schema<$Context["scalars"]>
          >,
          "id"
        >
    >
  >;
}

export interface BuilderMethodsRoot<$Context extends $$Utilities.Context> {
  query: RootTypeCustomNameForQueryMethods<$Context>;
}

export interface BuilderMethodsRootFn extends $$Utilities.TypeFunction {
  // @ts-expect-error parameter is Untyped.
  return: BuilderMethodsRoot<this["params"]>;
}
"
`;

exports[`schema with long type name 1`] = `
"import * as $$Data from "./data.js";
import * as $$Scalar from "./scalar.js";
import type { Schema as $ } from "graffle/utilities-for-generated";
import type * as $$Utilities from "graffle/utilities-for-generated";

export namespace Schema {
  //
  //
  //
  //
  //
  //
  // ==================================================================================================
  //                                                Root
  // ==================================================================================================
  //
  //
  //
  //
  //
  //

  //                                               Query
  // --------------------------------------------------------------------------------------------------
  //

  export interface Query {
    kind: "Object";
    name: "Query";
    fields: {
      __typename: Query.__typename;
      longTypeName: Query.longTypeName;
    };
  }

  export namespace Query {
    export interface __typename {
      kind: "OutputField";
      name: "__typename";
      arguments: {};
      inlineType: [1];
      namedType: {
        kind: "__typename";
        value: "Query";
      };
    }

    export interface longTypeName {
      kind: "OutputField";
      name: "longTypeName";
      arguments: {};
      inlineType: [0];
      namedType: $$NamedTypes.$$aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa;
    }
  }

  //
  //
  //
  //
  //
  //
  // ==================================================================================================
  //                                            OutputObject
  // ==================================================================================================
  //
  //
  //
  //
  //
  //

  // aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
  // --------------------------------------------------------------------------------------------------
  //

  export interface aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa {
    kind: "Object";
    name: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";
    fields: {
      __typename: aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa.__typename;
      id: aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa.id;
    };
  }

  export namespace aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa {
    export interface __typename {
      kind: "OutputField";
      name: "__typename";
      arguments: {};
      inlineType: [1];
      namedType: {
        kind: "__typename";
        value: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";
      };
    }

    export interface id {
      kind: "OutputField";
      name: "id";
      arguments: {};
      inlineType: [0];
      namedType: $$NamedTypes.$$ID;
    }
  }

  //
  //
  //
  //
  //
  //
  // ==================================================================================================
  //                                            InputObject
  // ==================================================================================================
  //
  //
  //
  //
  //
  //

  //
  //
  //
  //
  //
  //
  // ==================================================================================================
  //                                             Interface
  // ==================================================================================================
  //
  //
  //
  //
  //
  //

  //
  //
  //
  //
  //
  //
  // ==================================================================================================
  //                                               Union
  // ==================================================================================================
  //
  //
  //
  //
  //
  //

  //
  //
  //
  //
  //
  //
  // ==================================================================================================
  //                                                Enum
  // ==================================================================================================
  //
  //
  //
  //
  //
  //

  //
  //
  //
  //
  //
  //
  // ==================================================================================================
  //                                            ScalarCustom
  // ==================================================================================================
  //
  //
  //
  //
  //
  //

  //
  //
  //
  //
  //
  //
  // ==================================================================================================
  //                                           ScalarStandard
  // ==================================================================================================
  //
  //
  //
  //
  //
  //

  //                                                 ID
  // --------------------------------------------------------------------------------------------------
  //

  export type ID = $.StandardTypes.ID;

  //                                              Boolean
  // --------------------------------------------------------------------------------------------------
  //

  export type Boolean = $.StandardTypes.Boolean;

  //                                               String
  // --------------------------------------------------------------------------------------------------
  //

  export type String = $.StandardTypes.String;

  //
  //
  //
  //
  //
  //
  // ==================================================================================================
  //                                         Named Types Index
  // ==================================================================================================
  //
  //
  //
  //
  //
  //

  /**
   * [1] These definitions serve to allow field selection interfaces to extend their respective object type without
   *     name clashing between the field name and the object name.
   *
   *     For example imagine \`Query.Foo\` field with type also called \`Foo\`. Our generated interfaces for each field
   *     would end up with an error of \`export interface Foo extends Foo ...\`
   */

  namespace $$NamedTypes {
    export type $$Query = Query;
    export type $$aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa =
      aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa;
    export type $$ID = ID;
    export type $$Boolean = Boolean;
    export type $$String = String;
  }
}

//
//
//
//
//
//
// ==================================================================================================
//                                               Schema
// ==================================================================================================
//
//
//
//
//
//

export interface Schema<
  $Scalars extends $$Utilities.Schema.Scalar.Registry = $$Scalar.$Registry,
> {
  name: $$Data.Name;
  operationsAvailable: ["query"];
  RootUnion: Schema.Query;
  Root: {
    query: Schema.Query;
    mutation: null;
    subscription: null;
  };
  allTypes: {
    Query: Schema.Query;
    aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa: Schema.aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa;
  };
  objects: {
    aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa: Schema.aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa;
  };
  unions: {};
  interfaces: {};
  scalarNamesUnion: "ID" | "Boolean" | "String";
  scalars: {
    ID: Schema.ID;
    Boolean: Schema.Boolean;
    String: Schema.String;
  };
  scalarRegistry: $Scalars;
  extensions: $$Utilities.GlobalRegistry.TypeExtensions;
}
"
`;
